aten/src/ATen/core/type.cpp

```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
       "' but it contains an Any type. Any types cannot be members of modules, classes, or named tuples.");
 }
 
+namespace {
+bool isSpecialChar(char a) {
+  return a == '[' || a == ',' || a == ']';
+}
+
+class TypeParser {
+ public:
+  explicit TypeParser(const std::string& pythonStr)
+    : pythonStr_(pythonStr), start_(0) {}
+    
+  TypePtr parse() {
+    std::string token;
+    for (; start_ < pythonStr_.size() && !isSpecialChar(pythonStr_[start_]);
+           ++start_) {
+      if (pythonStr_[start_] != ' ') {
+        token += pythonStr_[start_];
```

Comment:
Char-by-char string creation is going to be n^2. This should only extract the string when it has found the end.

Reply:
The logic is changed to index points and integrated in lex() function.